src/middlewared/middlewared/plugins/reporting.py

```diff
@@ -0,0 +1,951 @@
+        if reg:
+            return (reg.group(1), int(reg.group(2)))
+
+    def get_identifiers(self):
+        return None
+
+    def encode(self, identifier):
+        return identifier
+
+    def get_defs(self, identifier):
+
+        rrd_types = self.get_rrd_types(identifier)
+        if not rrd_types:
+            raise RuntimeError(f'rrd_types not defined for {self.name!r}')
+
+        args = []
+        defs = {}
+        for i, rrd_type in enumerate(rrd_types):
+            _type, dsname, transform = rrd_type
+            direc = self.plugin
+            if self.identifier_plugin and identifier:
+                identifier = self.encode(identifier)
+                direc += f'-{identifier}'
+            path = os.path.join(self._base_path, direc, f'{_type}.rrd')
+            path = path.replace(':', r'\:')
+            name = f'{_type}_{dsname}'
+            defs[i] = {
+                'name': name,
+                'transform': transform,
+            }
+            args += [
+                f'DEF:{name}={path}:{dsname}:AVERAGE',
+            ]
+
+        for i, attrs in defs.items():
+            if attrs['transform']:
+                transform = attrs['transform']
+                if '%name%' in transform:
+                    transform = transform.replace('%name%', attrs['name'])
+                for orig, number in RE_NAME.findall(transform):
+                    transform = transform.replace(orig, defs[int(number)]['name'])
+                args += [
+                    f'CDEF:c{attrs["name"]}={transform}',
+                    f'XPORT:c{attrs["name"]}:{attrs["name"]}',
+                ]
+            else:
+                args += [f'XPORT:{attrs["name"]}:{attrs["name"]}']
+
+        if self.rrd_data_extra:
+            extra = textwrap.dedent(self.rrd_data_extra)
+            for orig, number in RE_NAME.findall(extra):
+                def_ = defs[int(number)]
+                name = def_['name']
+                if def_['transform']:
+                    name = 'c' + name
+                extra = extra.replace(orig, name)
+            args += extra.split()
+
+        return args
+
+    def export(self, identifier, starttime, endtime, aggregate=True):
+        args = [
+            'rrdtool',
+            'xport',
+            '--daemon', 'unix:/var/run/rrdcached.sock',
+            '--json',
+            '--end', endtime,
+            '--start', starttime,
+        ]
+        args.extend(self.get_defs(identifier))
+        cp = subprocess.run(args, capture_output=True)
+        if cp.returncode != 0:
+            raise RuntimeError(f'Failed to export RRD data: {cp.stderr.decode()}')
+
+        data = json.loads(cp.stdout)
+        data = dict(
+            name=self.name,
+            identifier=identifier,
+            data=data['data'],
+            **data['meta'],
+            aggregations=dict(),
+        )
+
+        if self.aggregations and aggregate:
+            df = pandas.DataFrame(data['data'])
```

Comment:
Isn't having `pandas` (and, thus, `numpy`) too heavy for us?

Reply:
It adds a few megabytes to the image, it seemed reasonable for the added performance improvement, doing it in pure python would be slower.

Do you have any proposal?

Reply:
A few megabytes is ok. In my memory it's around 50-100 for numpy and all it's shared libraries dependencies.

Reply:
I had not realized its such a big memory hog. Is there any alternative over doing it pure python?

Reply:
Sorry, I've meant in my human memory, when I was building linux docker images, they got bigger by 100 mb after apt-get install numpy. If FreeNAS ISO and RAM usage seem sane to you (I have no idea what they are), I am ok with that.

Reply:
Hmm yeah, it seems I miscalculated it, I had forgot to account for gcc7 which adds about 40MB to the image

Reply:
I will add a new ticket to find an alternative before we release 11.3